xdsl/rewriting/composable_rewriting/immutable_ir/immutable_ir.py

```diff
@@ -0,0 +1,550 @@
+        self.ops.freeze()
+
+    @classmethod
+    def from_mutable(
+        cls,
+        block: Block,
+        value_map: Optional[dict[SSAValue, ISSAValue]] = None,
+        block_map: Optional[dict[Block, IBlock]] = None,
+    ) -> IBlock:
+        """
+        Creates an immutable block from a mutable block. 
+        The value_map and block_map are used to map already known correspondings
+        of mutable values to immutable values and mutable blocks to immutable blocks.
+        """
+        if value_map is None:
+            value_map = {}
+        if block_map is None:
+            block_map = {}
+
+        args: List[IBlockArg] = []
+        for arg in block.args:
+            # The IBlock that will house this IBlockArg is not constructed yet.
+            # After construction the block field will be set by the IBlock.
+            immutable_arg = IBlockArg(arg.typ, IList([]), None, arg.index)  # type: ignore
+            args.append(immutable_arg)
+            value_map[arg] = immutable_arg
+
+        immutable_ops = [
+            IOp.from_mutable(op, value_map=value_map, block_map=block_map, existing_operands=None)
+            for op in block.ops
+        ]
+
+        return IBlock(args, immutable_ops)
+
+    def get_mutable_copy(
+            self,
+            value_mapping: Optional[dict[ISSAValue, SSAValue]] = None,
+            block_mapping: Optional[dict[IBlock, Block]] = None) -> Block:
+        """
+        Returns a mutable block that is a copy of this immutable block.
+        The value_mapping and block_mapping are used to map already known correspondings
+        of immutable values to mutable values and immutable blocks to mutable blocks.
+        """
+        if value_mapping is None:
+            value_mapping = {}
+        if block_mapping is None:
+            block_mapping = {}
+
+
+        # Block might already have been created by the Region, look it up
+        if self in block_mapping:
+            mutable_block = block_mapping[self]
+        else:
+            mutable_block = Block.from_arg_types(self.arg_types)
+        for idx, arg in enumerate(self.args):
+            value_mapping[arg] = mutable_block.args[idx]
+        block_mapping[self] = mutable_block
+
+        for immutable_op in self.ops:
+            mutable_block.add_op(
+                immutable_op.get_mutable_copy(value_mapping=value_mapping,
+                                              block_mapping=block_mapping))
+        return mutable_block
+
+
+def get_immutable_copy(op: Operation) -> IOp:
+    return IOp.from_mutable(op, {})
+
+
+@dataclass(frozen=True)
+class OpData:
+    """
+    Represents split off fields of IOp to its own class because they are
+    often preserved during rewriting. A new operation of the same type, e.g.
+    with changed operands can still use the same OpData instance. This design
+    increases sharing in the IR.
+    """
+    name: str
+    op_type: type[Operation]
+    attributes: dict[str, Attribute]
```

Comment:
Is there a reason why this is mutable?
Should we have an `IDict`?

Reply:
Right, this should not be mutable. For Python there already exists a `frozendict`, which we could pick up as a dependency. 
I opted to reimplement the `frozenlist` to achieve nicer pattern matching. I have not experimented with how pattern matching looks like with `frozendict`

Reply:
I moved the implementation to `frozendict` but sadly it seems it does not properly support type annotations.